In [2]:

import pandas as pd

In [3]:
df = pd.read_csv("C:\\Users\\Sandeep\\OneDrive\\Documents\\Projects\\IPL-Performance-Intelligence\\data\\raw\\IPL.csv")

C:\Users\Sandeep\AppData\Local\Temp\ipykernel_24408\758258734.py:1: DtypeWarning: Columns (0: review_batter, 1: team_reviewed, 2: review_decision, 3: umpire, 4: season, 5: superover_winner, 6: result_type, 7: method, 8: event_match_no) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv("C:\\Users\\Sandeep\\OneDrive\\Documents\\Projects\\IPL-Performance-Intelligence\\data\\raw\\IPL.csv")


In [4]:
df.head()

,match_id,date,match_type,event_name,innings,batting_team,bowling_team,over,ball,ball_no,...,team_balls,team_wicket,new_batter,power_surge_start,batter_runs,batter_balls,bowler_wicket,batting_partners,striker_out,next_batter
0,335982,2008-04-18,T20,Indian Premier League,1,Kolkata Knight Riders,Royal Challengers Bangalore,0,1,0.1,...,1,0,NaN,NaN,0,1,0,BB McCullum|SC Ganguly,0,NaN
1,335982,2008-04-18,T20,Indian Premier League,1,Kolkata Knight Riders,Royal Challengers Bangalore,0,2,0.2,...,2,0,NaN,NaN,0,1,0,BB McCullum|SC Ganguly,0,NaN
2,335982,2008-04-18,T20,Indian Premier League,1,Kolkata Knight Riders,Royal Challengers Bangalore,0,3,0.3,...,2,0,NaN,NaN,0,1,0,BB McCullum|SC Ganguly,0,NaN
3,335982,2008-04-18,T20,Indian Premier League,1,Kolkata Knight Riders,Royal Challengers Bangalore,0,3,0.3,...,3,0,NaN,NaN,0,2,0,BB McCullum|SC Ganguly,0,NaN
4,335982,2008-04-18,T20,Indian Premier League,1,Kolkata Knight Riders,Royal Challengers Bangalore,0,4,0.4,...,4,0,NaN,NaN,0,3,0,BB McCullum|SC Ganguly,0,NaN


In [5]:
df.shape

(295732, 64)

In [6]:
df.columns

Index(['match_id', 'date', 'match_type', 'event_name', 'innings',
       'batting_team', 'bowling_team', 'over', 'ball', 'ball_no', 'batter',
       'bat_pos', 'runs_batter', 'balls_faced', 'bowler', 'valid_ball',
       'runs_extras', 'runs_total', 'runs_bowler', 'runs_not_boundary',
       'extra_type', 'non_striker', 'non_striker_pos', 'wicket_kind',
       'player_out', 'fielders', 'runs_target', 'review_batter',
       'team_reviewed', 'review_decision', 'umpire', 'umpires_call',
       'player_of_match', 'match_won_by', 'win_outcome', 'toss_winner',
       'toss_decision', 'venue', 'city', 'day', 'month', 'year', 'season',
       'gender', 'team_type', 'superover_winner', 'result_type', 'method',
       'balls_per_over', 'overs', 'event_match_no', 'stage', 'match_number',
       'team_runs', 'team_balls', 'team_wicket', 'new_batter',
       'power_surge_start', 'batter_runs', 'batter_balls', 'bowler_wicket',
       'batting_partners', 'striker_out', 'next_batter'],
      dtype='s

In [7]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 295732 entries, 0 to 295731
Data columns (total 64 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   match_id           295732 non-null  int64  
 1   date               295732 non-null  str    
 2   match_type         295732 non-null  str    
 3   event_name         295732 non-null  str    
 4   innings            295732 non-null  int64  
 5   batting_team       295732 non-null  str    
 6   bowling_team       295732 non-null  str    
 7   over               295732 non-null  int64  
 8   ball               295732 non-null  int64  
 9   ball_no            295732 non-null  float64
 10  batter             295732 non-null  str    
 11  bat_pos            295732 non-null  int64  
 12  runs_batter        295732 non-null  int64  
 13  balls_faced        295732 non-null  int64  
 14  bowler             295732 non-null  str    
 15  valid_ball         295732 non-null  int64  
 16  runs_extras  

In [8]:
df.describe()

,match_id,innings,over,ball,ball_no,bat_pos,runs_batter,balls_faced,valid_ball,runs_extras,...,balls_per_over,overs,team_runs,team_balls,team_wicket,power_surge_start,batter_runs,batter_balls,bowler_wicket,striker_out
count,2.957320e+05,295732.000000,295732.000000,295732.000000,295732.000000,295732.000000,295732.000000,295732.000000,295732.000000,295732.000000,...,295732.0,295732.0,295732.000000,295732.000000,295732.000000,0.0,295732.000000,295732.000000,295732.000000,295732.000000
mean,9.770551e+05,1.482728,9.190331,3.488807,9.539212,3.621847,1.290117,0.966605,0.962459,0.068336,...,6.0,20.0,78.146291,58.592844,2.466067,NaN,18.514882,13.982883,0.045609,0.047793
std,3.953255e+05,0.502436,5.679690,1.708185,5.681079,2.173463,1.664418,0.179666,0.190083,0.342939,...,0.0,0.0,50.698248,34.106127,2.105231,NaN,18.767165,11.811388,0.208636,0.213329
min,3.359820e+05,1.000000,0.000000,1.000000,0.100000,1.000000,0.000000,0.000000,0.000000,0.000000,...,6.0,20.0,0.000000,0.000000,0.000000,NaN,0.000000,0.000000,0.000000,0.000000
25%,5.483720e+05,1.000000,4.000000,2.000000,4.500000,2.000000,0.000000,1.000000,1.000000,0.000000,...,6.0,20.0,36.000000,29.000000,1.000000,NaN,4.000000,5.000000,0.000000,0.000000
50%,1.082639e+06,1.000000,9.000000,3.000000,9.400000,3.000000,1.000000,1.000000,1.000000,0.000000,...,6.0,20.0,74.000000,58.000000,2.000000,NaN,13.000000,11.000000,0.000000,0.000000
75%,1.304103e+06,2.000000,14.000000,5.000000,14.400000,5.000000,1.000000,1.000000,1.000000,0.000000,...,6.0,20.0,115.000000,88.000000,4.000000,NaN,27.000000,20.000000,0.000000,0.000000
max,1.535465e+06,6.000000,19.000000,7.000000,19.600000,11.000000,6.000000,1.000000,1.000000,7.000000,...,6.0,20.0,287.000000,121.000000,10.000000,NaN,175.000000,73.000000,1.000000,1.000000


In [9]:
pd.set_option('display.max_columns',None)

In [10]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 295732 entries, 0 to 295731
Data columns (total 64 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   match_id           295732 non-null  int64  
 1   date               295732 non-null  str    
 2   match_type         295732 non-null  str    
 3   event_name         295732 non-null  str    
 4   innings            295732 non-null  int64  
 5   batting_team       295732 non-null  str    
 6   bowling_team       295732 non-null  str    
 7   over               295732 non-null  int64  
 8   ball               295732 non-null  int64  
 9   ball_no            295732 non-null  float64
 10  batter             295732 non-null  str    
 11  bat_pos            295732 non-null  int64  
 12  runs_batter        295732 non-null  int64  
 13  balls_faced        295732 non-null  int64  
 14  bowler             295732 non-null  str    
 15  valid_ball         295732 non-null  int64  
 16  runs_extras  

In [11]:
df['match_type'].unique()

<StringArray>
['T20']
Length: 1, dtype: str

In [12]:
df['season'].unique()

array(['2007/08', '2009', 2009, '2009/10', '2011', 2011, 2012, 2013, 2014,
       2015, 2016, 2017, 2018, 2019, '2019', '2020/21', '2021', 2021,
       2022, 2023, 2024, 2025, 2026], dtype=object)

In [13]:
df['batting_team'].unique()

<StringArray>
[      'Kolkata Knight Riders', 'Royal Challengers Bangalore',
         'Chennai Super Kings',             'Kings XI Punjab',
            'Rajasthan Royals',            'Delhi Daredevils',
              'Mumbai Indians',             'Deccan Chargers',
        'Kochi Tuskers Kerala',               'Pune Warriors',
         'Sunrisers Hyderabad',     'Rising Pune Supergiants',
               'Gujarat Lions',      'Rising Pune Supergiant',
              'Delhi Capitals',                'Punjab Kings',
        'Lucknow Super Giants',              'Gujarat Titans',
 'Royal Challengers Bengaluru']
Length: 19, dtype: str

In [14]:
df['batter'].nunique()

738

In [15]:
df['bowler'].nunique()

577

In [16]:
df.isnull().sum()

match_id                 0
date                     0
match_type               0
event_name               0
innings                  0
                     ...  
batter_balls             0
bowler_wicket            0
batting_partners         0
striker_out              0
next_batter         281569
Length: 64, dtype: int64

In [17]:
df.isnull().sum().sort_values(ascending=True)

match_id                  0
date                      0
match_type                0
event_name                0
innings                   0
                      ...  
review_batter        294742
umpire               294742
review_decision      294742
team_reviewed        294742
power_surge_start    295732
Length: 64, dtype: int64

In [18]:
missing_percentage = (df.isnull().sum()/len(df)) * 100
missing_percentage.sort_values(ascending=False)

power_surge_start    100.000000
review_batter         99.665237
review_decision       99.665237
umpire                99.665237
team_reviewed         99.665237
                        ...    
batter_runs            0.000000
batter_balls           0.000000
bowler_wicket          0.000000
batting_partners       0.000000
striker_out            0.000000
Length: 64, dtype: float64

In [19]:
missing = df.isnull().sum()
missing[missing>0].sort_values(ascending=False)

power_surge_start    295732
team_reviewed        294742
review_batter        294742
review_decision      294742
umpire               294742
method               291601
superover_winner     291580
result_type          290752
fielders             285011
new_batter           281569
next_batter          281569
wicket_kind          281027
player_out           281027
extra_type           279505
runs_target          153429
win_outcome            4980
dtype: int64

In [20]:
missing_percentage2 = (missing[missing>0]/len(df))*100
missing_percentage2.sort_values(ascending=False)

power_surge_start    100.000000
team_reviewed         99.665237
review_batter         99.665237
review_decision       99.665237
umpire                99.665237
method                98.603127
superover_winner      98.596026
result_type           98.316043
fielders              96.374758
new_batter            95.210867
next_batter           95.210867
wicket_kind           95.027593
player_out            95.027593
extra_type            94.512937
runs_target           51.881095
win_outcome            1.683957
dtype: float64

In [21]:
columns_to_drop = [
    'power_surge_start',
    'team_reviewed',
    'review_batter',
    'review_decision',
    'umpire',
    'new_batter',
    'next_batter'
]
df = df.drop(columns = columns_to_drop)

In [22]:
df.isnull().sum().sort_values(ascending=False)

method               291601
superover_winner     291580
result_type          290752
fielders             285011
player_out           281027
wicket_kind          281027
extra_type           279505
runs_target          153429
win_outcome            4980
ball                      0
over                      0
bowling_team              0
ball_no                   0
balls_faced               0
batter                    0
bat_pos                   0
runs_batter               0
runs_extras               0
valid_ball                0
date                      0
innings                   0
batting_team              0
match_type                0
event_name                0
match_id                  0
non_striker               0
non_striker_pos           0
runs_bowler               0
runs_not_boundary         0
runs_total                0
bowler                    0
player_of_match           0
umpires_call              0
venue                     0
city                      0
day                 

In [23]:
df['win_outcome'].unique

<bound method Series.unique of 0          140 runs
1          140 runs
2          140 runs
3          140 runs
4          140 runs
            ...    
295727    5 wickets
295728    5 wickets
295729    5 wickets
295730    5 wickets
295731    5 wickets
Name: win_outcome, Length: 295732, dtype: str>

In [24]:
df['win_outcome'].value_counts(dropna=False)

win_outcome
6 wickets    33237
7 wickets    30455
5 wickets    26905
8 wickets    20523
4 wickets    16021
             ...  
89 runs        215
138 runs       212
146 runs       211
76 runs        204
112 runs       189
Name: count, Length: 115, dtype: int64

In [25]:
df[df['win_outcome'].isnull()][
    ['match_id', 'season', 'batting_team', 'bowling_team', 'match_won_by', 'win_outcome']
].head(20)

,match_id,season,batting_team,bowling_team,match_won_by,win_outcome
15161,392190,2009,Rajasthan Royals,Kolkata Knight Riders,Unknown,NaN
15162,392190,2009,Rajasthan Royals,Kolkata Knight Riders,Unknown,NaN
15163,392190,2009,Rajasthan Royals,Kolkata Knight Riders,Unknown,NaN
15164,392190,2009,Rajasthan Royals,Kolkata Knight Riders,Unknown,NaN
15165,392190,2009,Rajasthan Royals,Kolkata Knight Riders,Unknown,NaN
15166,392190,2009,Rajasthan Royals,Kolkata Knight Riders,Unknown,NaN
15167,392190,2009,Rajasthan Royals,Kolkata Knight Riders,Unknown,NaN
15168,392190,2009,Rajasthan Royals,Kolkata Knight Riders,Unknown,NaN
15169,392190,2009,Rajasthan Royals,Kolkata Knight Riders,Unknown,NaN
15170,392190,2009,Rajasthan Royals,Kolkata Knight Riders,Unknown,NaN


In [26]:
df[df['win_outcome'].isnull()][
    ['match_id', 'season', 'batting_team', 'bowling_team',
     'match_won_by', 'win_outcome']
].drop_duplicates().head(20)

,match_id,season,batting_team,bowling_team,match_won_by,win_outcome
15161,392190,2009,Rajasthan Royals,Kolkata Knight Riders,Unknown,NaN
15287,392190,2009,Kolkata Knight Riders,Rajasthan Royals,Unknown,NaN
30707,419121,2009/10,Kings XI Punjab,Chennai Super Kings,Unknown,NaN
30829,419121,2009/10,Chennai Super Kings,Kings XI Punjab,Unknown,NaN
57068,501265,2011,Delhi Daredevils,Pune Warriors,Unknown,NaN
77794,598004,2013,Royal Challengers Bangalore,Sunrisers Hyderabad,Unknown,NaN
77917,598004,2013,Sunrisers Hyderabad,Royal Challengers Bangalore,Unknown,NaN
81200,598017,2013,Delhi Daredevils,Royal Challengers Bangalore,Unknown,NaN
81324,598017,2013,Royal Challengers Bangalore,Delhi Daredevils,Unknown,NaN
98800,729315,2014,Rajasthan Royals,Kolkata Knight Riders,Unknown,NaN


In [27]:
df[df['win_outcome'].isnull()]['match_id'].nunique()

25

In [28]:
df.duplicated().sum()

np.int64(0)

In [29]:
df.duplicated(
    subset=['match_id', 'innings', 'over', 'ball'],
    keep=False
).sum()

np.int64(21493)

In [30]:
duplicate_balls = df[df.duplicated(subset = ['match_id','innings','over','ball'], keep = False)]

duplicate_balls[['match_id', 'innings', 'over', 'ball',
     'batter', 'bowler', 'runs_batter',
     'runs_extras', 'runs_total', 'extra_type']].head(20)

,match_id,innings,over,ball,batter,bowler,runs_batter,runs_extras,runs_total,extra_type
2,335982,1,0,3,BB McCullum,P Kumar,0,1,1,wides
3,335982,1,0,3,BB McCullum,P Kumar,0,0,0,NaN
19,335982,1,3,1,BB McCullum,AA Noffke,0,5,5,wides
20,335982,1,3,1,BB McCullum,AA Noffke,6,0,6,NaN
78,335982,1,12,5,BB McCullum,JH Kallis,0,1,1,wides
79,335982,1,12,5,BB McCullum,JH Kallis,4,0,4,NaN
91,335982,1,14,5,BB McCullum,CL White,0,2,2,wides
92,335982,1,14,5,DJ Hussey,CL White,1,0,1,NaN
125,335982,2,0,2,W Jaffer,AB Dinda,0,1,1,wides
126,335982,2,0,2,W Jaffer,AB Dinda,0,0,0,NaN


In [31]:
df.dtypes

match_id               int64
date                     str
match_type               str
event_name               str
innings                int64
batting_team             str
bowling_team             str
over                   int64
ball                   int64
ball_no              float64
batter                   str
bat_pos                int64
runs_batter            int64
balls_faced            int64
bowler                   str
valid_ball             int64
runs_extras            int64
runs_total             int64
runs_bowler            int64
runs_not_boundary       bool
extra_type               str
non_striker              str
non_striker_pos        int64
wicket_kind              str
player_out               str
fielders                 str
runs_target          float64
umpires_call            bool
player_of_match          str
match_won_by             str
win_outcome              str
toss_winner              str
toss_decision            str
venue                    str
city          

In [32]:
df['season'].value_counts().sort_index

<bound method Series.sort_index of season
2013       18177
2022       17912
2023       17863
2012       17767
2026       17527
2025       17285
2024       17103
2020/21    14559
2009/10    14498
2014       14300
2018       14286
2016       14096
2017       13862
2015       13652
2007/08    13489
2021       11434
2011        9454
2009        8192
2011        7559
2019        7286
2019        7026
2009        5414
2021        2991
Name: count, dtype: int64>

In [33]:
print(df['batting_team'].unique())

<StringArray>
[      'Kolkata Knight Riders', 'Royal Challengers Bangalore',
         'Chennai Super Kings',             'Kings XI Punjab',
            'Rajasthan Royals',            'Delhi Daredevils',
              'Mumbai Indians',             'Deccan Chargers',
        'Kochi Tuskers Kerala',               'Pune Warriors',
         'Sunrisers Hyderabad',     'Rising Pune Supergiants',
               'Gujarat Lions',      'Rising Pune Supergiant',
              'Delhi Capitals',                'Punjab Kings',
        'Lucknow Super Giants',              'Gujarat Titans',
 'Royal Challengers Bengaluru']
Length: 19, dtype: str


In [34]:
print(df['bowling_team'].unique())

<StringArray>
['Royal Challengers Bangalore',       'Kolkata Knight Riders',
             'Kings XI Punjab',         'Chennai Super Kings',
            'Delhi Daredevils',            'Rajasthan Royals',
              'Mumbai Indians',             'Deccan Chargers',
        'Kochi Tuskers Kerala',               'Pune Warriors',
         'Sunrisers Hyderabad',     'Rising Pune Supergiants',
               'Gujarat Lions',      'Rising Pune Supergiant',
              'Delhi Capitals',                'Punjab Kings',
              'Gujarat Titans',        'Lucknow Super Giants',
 'Royal Challengers Bengaluru']
Length: 19, dtype: str


In [35]:
df['season'].apply(type).value_counts()

season
<class 'int'>    230196
<class 'str'>     65536
Name: count, dtype: int64

In [36]:
for value in df['season'].unique():
    print(value, "→", type(value))

2007/08 → <class 'str'>
2009 → <class 'str'>
2009 → <class 'int'>
2009/10 → <class 'str'>
2011 → <class 'str'>
2011 → <class 'int'>
2012 → <class 'int'>
2013 → <class 'int'>
2014 → <class 'int'>
2015 → <class 'int'>
2016 → <class 'int'>
2017 → <class 'int'>
2018 → <class 'int'>
2019 → <class 'int'>
2019 → <class 'str'>
2020/21 → <class 'str'>
2021 → <class 'str'>
2021 → <class 'int'>
2022 → <class 'int'>
2023 → <class 'int'>
2024 → <class 'int'>
2025 → <class 'int'>
2026 → <class 'int'>


In [37]:
df['season'].astype(str).value_counts().sort_index()

season
2007/08    13489
2009       13606
2009/10    14498
2011       17013
2012       17767
2013       18177
2014       14300
2015       13652
2016       14096
2017       13862
2018       14286
2019       14312
2020/21    14559
2021       14425
2022       17912
2023       17863
2024       17103
2025       17285
2026       17527
Name: count, dtype: int64

In [38]:
# Check team names and their number of occurrences
df['batting_team'].value_counts()

batting_team
Mumbai Indians                 34990
Kolkata Knight Riders          32463
Chennai Super Kings            32101
Rajasthan Royals               29777
Royal Challengers Bangalore    28205
Sunrisers Hyderabad            25284
Kings XI Punjab                22646
Delhi Daredevils               18786
Delhi Capitals                 14321
Punjab Kings                   10454
Gujarat Titans                  9379
Deccan Chargers                 9034
Lucknow Super Giants            8854
Pune Warriors                   5443
Royal Challengers Bengaluru     5367
Gujarat Lions                   3566
Rising Pune Supergiant          1900
Kochi Tuskers Kerala            1582
Rising Pune Supergiants         1580
Name: count, dtype: int64

In [39]:
# Check team name differences between batting and bowling teams
set(df['batting_team'].unique()) ^ set(df['bowling_team'].unique())

set()

In [40]:
# Standardize inconsistent team name
df['batting_team'] = df['batting_team'].replace(
    'Rising Pune Supergiants',
    'Rising Pune Supergiant'
)

df['bowling_team'] = df['bowling_team'].replace(
    'Rising Pune Supergiants',
    'Rising Pune Supergiant'
)

In [41]:
df['batting_team'].value_counts()

batting_team
Mumbai Indians                 34990
Kolkata Knight Riders          32463
Chennai Super Kings            32101
Rajasthan Royals               29777
Royal Challengers Bangalore    28205
Sunrisers Hyderabad            25284
Kings XI Punjab                22646
Delhi Daredevils               18786
Delhi Capitals                 14321
Punjab Kings                   10454
Gujarat Titans                  9379
Deccan Chargers                 9034
Lucknow Super Giants            8854
Pune Warriors                   5443
Royal Challengers Bengaluru     5367
Gujarat Lions                   3566
Rising Pune Supergiant          3480
Kochi Tuskers Kerala            1582
Name: count, dtype: int64

In [42]:
# Check unique values in important categorical columns
categorical_cols = [
    'match_type',
    'event_name',
    'extra_type',
    'wicket_kind',
    'match_won_by',
    'win_outcome',
    'toss_decision',
    'venue',
    'city',
    'gender',
    'team_type',
    'result_type',
    'method',
    'stage'
]

for col in categorical_cols:
    print(f"\n{'='*50}")
    print(f"{col.upper()} ({df[col].nunique()} unique values)")
    print('='*50)
    print(df[col].value_counts(dropna=False).head(30))


MATCH_TYPE (1 unique values)
match_type
T20    295732
Name: count, dtype: int64

EVENT_NAME (1 unique values)
event_name
Indian Premier League    295732
Name: count, dtype: int64

EXTRA_TYPE (8 unique values)
extra_type
NaN                 279505
wides                 9875
legbyes               4394
noballs               1199
byes                   730
byes, noballs           15
legbyes, noballs        12
penalty                  1
penalty, wides           1
Name: count, dtype: int64

WICKET_KIND (10 unique values)
wicket_kind
NaN                      281027
caught                     9321
bowled                     2460
run out                    1194
lbw                         883
caught and bowled           410
stumped                     388
hit wicket                   20
retired hurt                 19
retired out                   6
obstructing the field         4
Name: count, dtype: int64

MATCH_WON_BY (20 unique values)
match_won_by
Mumbai Indians                 37151
Chenn

In [43]:
df['match_won_by'] = df['match_won_by'].replace(
    'Rising Pune Supergiants',
    'Rising Pune Supergiant'
)

In [44]:
df['city'] = df['city'].replace(
    'Bangalore',
    'Bengaluru'
)

In [45]:
print(df['match_won_by'].value_counts())
print("\nCity counts:")
print(df['city'].value_counts())

match_won_by
Mumbai Indians                 37151
Chennai Super Kings            35839
Kolkata Knight Riders          32921
Rajasthan Royals               29528
Royal Challengers Bangalore    26801
Sunrisers Hyderabad            24431
Kings XI Punjab                20162
Delhi Daredevils               15709
Delhi Capitals                 14008
Gujarat Titans                 11339
Punjab Kings                    9887
Lucknow Super Giants            8396
Deccan Chargers                 7013
Royal Challengers Bengaluru     6811
Unknown                         4980
Rising Pune Supergiant          3488
Gujarat Lions                   3063
Pune Warriors                   2883
Kochi Tuskers Kerala            1322
Name: count, dtype: int64

City counts:
city
Mumbai            45107
Kolkata           24936
Delhi             24566
Bengaluru         24128
Chennai           23851
Hyderabad         21416
Jaipur            16174
Chandigarh        14476
Ahmedabad         12671
Unknown           12397

In [46]:
# Find all columns containing "Royal Challengers"
for col in df.columns:
    if df[col].dtype == 'object' or df[col].dtype == 'str':
        matches = df[col].astype(str).str.contains(
            'Royal Challengers',
            case=False,
            na=False
        ).sum()
        
        if matches > 0:
            print(f"{col}: {matches}")
            

batting_team: 33572
bowling_team: 33827
match_won_by: 33612
toss_winner: 32017
superover_winner: 518


In [47]:
team_name_checks = [
    'Royal Challengers',
    'Rising Pune Supergiants'
]

for team_name in team_name_checks:
    print(f"\n{'='*60}")
    print(f"SEARCHING FOR: {team_name}")
    print('='*60)

    for col in df.columns:
        matches = df[col].astype(str).str.contains(
            team_name,
            case=False,
            na=False
        ).sum()

        if matches > 0:
            print(f"{col}: {matches}")


SEARCHING FOR: Royal Challengers
batting_team: 33572
bowling_team: 33827
match_won_by: 33612
toss_winner: 32017
superover_winner: 518

SEARCHING FOR: Rising Pune Supergiants
toss_winner: 1503


In [48]:
team_name_mapping = {
    'Royal Challengers Bangalore': 'Royal Challengers Bengaluru',
    'Rising Pune Supergiants': 'Rising Pune Supergiant'
}

In [49]:
team_columns = [
    'batting_team',
    'bowling_team',
    'match_won_by',
    'toss_winner',
    'superover_winner'
]

for col in team_columns:
    df[col] = df[col].replace(team_name_mapping)

In [50]:
df['batting_team'].value_counts()

batting_team
Mumbai Indians                 34990
Royal Challengers Bengaluru    33572
Kolkata Knight Riders          32463
Chennai Super Kings            32101
Rajasthan Royals               29777
Sunrisers Hyderabad            25284
Kings XI Punjab                22646
Delhi Daredevils               18786
Delhi Capitals                 14321
Punjab Kings                   10454
Gujarat Titans                  9379
Deccan Chargers                 9034
Lucknow Super Giants            8854
Pune Warriors                   5443
Gujarat Lions                   3566
Rising Pune Supergiant          3480
Kochi Tuskers Kerala            1582
Name: count, dtype: int64

In [51]:
team_columns = [
    'batting_team',
    'bowling_team',
    'match_won_by',
    'toss_winner',
    'superover_winner'
]

for col in team_columns:
    print(f"\n{col.upper()}")
    print("-" * 40)
    
    print(
        df[col].astype(str).str.contains(
            'Royal Challengers Bangalore|Rising Pune Supergiants',
            case=False,
            na=False
        ).sum()
    )


BATTING_TEAM
----------------------------------------
0

BOWLING_TEAM
----------------------------------------
0

MATCH_WON_BY
----------------------------------------
0

TOSS_WINNER
----------------------------------------
0

SUPEROVER_WINNER
----------------------------------------
0


In [52]:
team_name_mapping = {
    'Royal Challengers Bangalore': 'Royal Challengers Bengaluru',
    'Delhi Daredevils': 'Delhi Capitals',
    'Kings XI Punjab': 'Punjab Kings',
    'Rising Pune Supergiants': 'Rising Pune Supergiant'
}

In [53]:
team_columns = [
    'batting_team',
    'bowling_team',
    'match_won_by',
    'toss_winner',
    'superover_winner'
]

In [54]:
for col in team_columns:
    df[col] = df[col].replace(team_name_mapping)

In [55]:
print(df['batting_team'].unique())

<StringArray>
[      'Kolkata Knight Riders', 'Royal Challengers Bengaluru',
         'Chennai Super Kings',                'Punjab Kings',
            'Rajasthan Royals',              'Delhi Capitals',
              'Mumbai Indians',             'Deccan Chargers',
        'Kochi Tuskers Kerala',               'Pune Warriors',
         'Sunrisers Hyderabad',      'Rising Pune Supergiant',
               'Gujarat Lions',        'Lucknow Super Giants',
              'Gujarat Titans']
Length: 15, dtype: str


In [56]:
for col in team_columns:
    print(f"\n{col.upper()} - {df[col].nunique()} unique values")
    print(sorted(df[col].dropna().unique()))


BATTING_TEAM - 15 unique values
['Chennai Super Kings', 'Deccan Chargers', 'Delhi Capitals', 'Gujarat Lions', 'Gujarat Titans', 'Kochi Tuskers Kerala', 'Kolkata Knight Riders', 'Lucknow Super Giants', 'Mumbai Indians', 'Pune Warriors', 'Punjab Kings', 'Rajasthan Royals', 'Rising Pune Supergiant', 'Royal Challengers Bengaluru', 'Sunrisers Hyderabad']

BOWLING_TEAM - 15 unique values
['Chennai Super Kings', 'Deccan Chargers', 'Delhi Capitals', 'Gujarat Lions', 'Gujarat Titans', 'Kochi Tuskers Kerala', 'Kolkata Knight Riders', 'Lucknow Super Giants', 'Mumbai Indians', 'Pune Warriors', 'Punjab Kings', 'Rajasthan Royals', 'Rising Pune Supergiant', 'Royal Challengers Bengaluru', 'Sunrisers Hyderabad']

MATCH_WON_BY - 16 unique values
['Chennai Super Kings', 'Deccan Chargers', 'Delhi Capitals', 'Gujarat Lions', 'Gujarat Titans', 'Kochi Tuskers Kerala', 'Kolkata Knight Riders', 'Lucknow Super Giants', 'Mumbai Indians', 'Pune Warriors', 'Punjab Kings', 'Rajasthan Royals', 'Rising Pune Supergia

In [57]:
print("Number of unique venues:", df['venue'].nunique())

print("\nVENUE VALUES:")
for venue in sorted(df['venue'].dropna().unique()):
    print(venue)

Number of unique venues: 60

VENUE VALUES:
Arun Jaitley Stadium
Arun Jaitley Stadium, Delhi
Barabati Stadium
Barsapara Cricket Stadium, Guwahati
Bharat Ratna Shri Atal Bihari Vajpayee Ekana Cricket Stadium, Lucknow
Brabourne Stadium
Brabourne Stadium, Mumbai
Buffalo Park
De Beers Diamond Oval
Dr DY Patil Sports Academy
Dr DY Patil Sports Academy, Mumbai
Dr. Y.S. Rajasekhara Reddy ACA-VDCA Cricket Stadium
Dr. Y.S. Rajasekhara Reddy ACA-VDCA Cricket Stadium, Visakhapatnam
Dubai International Cricket Stadium
Eden Gardens
Eden Gardens, Kolkata
Feroz Shah Kotla
Green Park
Himachal Pradesh Cricket Association Stadium
Himachal Pradesh Cricket Association Stadium, Dharamsala
Holkar Cricket Stadium
JSCA International Stadium Complex
Kingsmead
M Chinnaswamy Stadium
M Chinnaswamy Stadium, Bengaluru
M.Chinnaswamy Stadium
MA Chidambaram Stadium
MA Chidambaram Stadium, Chepauk
MA Chidambaram Stadium, Chepauk, Chennai
Maharaja Yadavindra Singh International Cricket Stadium, Mullanpur
Maharaja Yadavin

In [58]:
df.groupby('venue')['city'].value_counts()

venue                                                                  city     
Arun Jaitley Stadium                                                   Delhi         3356
Arun Jaitley Stadium, Delhi                                            Delhi         7260
Barabati Stadium                                                       Cuttack       1695
Barsapara Cricket Stadium, Guwahati                                    Guwahati      1799
Bharat Ratna Shri Atal Bihari Vajpayee Ekana Cricket Stadium, Lucknow  Lucknow       6904
                                                                                    ...  
SuperSport Park                                                        Centurion     2866
Vidarbha Cricket Association Stadium, Jamtha                           Nagpur         742
Wankhede Stadium                                                       Mumbai       17591
Wankhede Stadium, Mumbai                                               Mumbai       14280
Zayed Cricket Stadi

In [59]:
venue_checks = [
    'Punjab Cricket Association',
    'Zayed',
    'Sardar Patel',
    'Narendra Modi'
]

for keyword in venue_checks:
    print(f"\n{'='*60}")
    print(f"SEARCHING: {keyword}")
    print('='*60)

    matches = df[
        df['venue'].astype(str).str.contains(
            keyword,
            case=False,
            na=False
        )
    ][['venue', 'city']].drop_duplicates()

    print(matches.to_string(index=False))


SEARCHING: Punjab Cricket Association
                                                           venue       city
                      Punjab Cricket Association Stadium, Mohali Chandigarh
            Punjab Cricket Association IS Bindra Stadium, Mohali Chandigarh
                    Punjab Cricket Association IS Bindra Stadium Chandigarh
Punjab Cricket Association IS Bindra Stadium, Mohali, Chandigarh Chandigarh

SEARCHING: Zayed
                           venue      city
            Sheikh Zayed Stadium Abu Dhabi
Zayed Cricket Stadium, Abu Dhabi Abu Dhabi

SEARCHING: Sardar Patel
                       venue      city
Sardar Patel Stadium, Motera Ahmedabad

SEARCHING: Narendra Modi
                           venue      city
Narendra Modi Stadium, Ahmedabad Ahmedabad


In [60]:
venue_mapping = {
    # Arun Jaitley
    'Arun Jaitley Stadium, Delhi': 'Arun Jaitley Stadium',

    # Brabourne
    'Brabourne Stadium, Mumbai': 'Brabourne Stadium',

    # Dr DY Patil
    'Dr DY Patil Sports Academy, Mumbai': 'Dr DY Patil Sports Academy',

    # ACA-VDCA
    'Dr. Y.S. Rajasekhara Reddy ACA-VDCA Cricket Stadium, Visakhapatnam':
        'Dr. Y.S. Rajasekhara Reddy ACA-VDCA Cricket Stadium',

    # Eden Gardens
    'Eden Gardens, Kolkata': 'Eden Gardens',

    # HPCA Stadium
    'Himachal Pradesh Cricket Association Stadium, Dharamsala':
        'Himachal Pradesh Cricket Association Stadium',

    # M Chinnaswamy
    'M Chinnaswamy Stadium, Bengaluru': 'M Chinnaswamy Stadium',
    'M.Chinnaswamy Stadium': 'M Chinnaswamy Stadium',

    # MA Chidambaram
    'MA Chidambaram Stadium, Chepauk': 'MA Chidambaram Stadium',
    'MA Chidambaram Stadium, Chepauk, Chennai': 'MA Chidambaram Stadium',

    # Maharaja Yadavindra Singh
    'Maharaja Yadavindra Singh International Cricket Stadium, Mullanpur':
        'Maharaja Yadavindra Singh International Cricket Stadium',
    'Maharaja Yadavindra Singh International Cricket Stadium, New Chandigarh':
        'Maharaja Yadavindra Singh International Cricket Stadium',

    # Maharashtra Cricket Association
    'Maharashtra Cricket Association Stadium, Pune':
        'Maharashtra Cricket Association Stadium',

    # Punjab Cricket Association
    'Punjab Cricket Association Stadium, Mohali':
        'Punjab Cricket Association IS Bindra Stadium',
    'Punjab Cricket Association IS Bindra Stadium, Mohali':
        'Punjab Cricket Association IS Bindra Stadium',
    'Punjab Cricket Association IS Bindra Stadium, Mohali, Chandigarh':
        'Punjab Cricket Association IS Bindra Stadium',

    # Rajiv Gandhi International Stadium
    'Rajiv Gandhi International Stadium, Uppal':
        'Rajiv Gandhi International Stadium',
    'Rajiv Gandhi International Stadium, Uppal, Hyderabad':
        'Rajiv Gandhi International Stadium',

    # Sawai Mansingh
    'Sawai Mansingh Stadium, Jaipur': 'Sawai Mansingh Stadium',

    # Shaheed Veer Narayan Singh
    'Shaheed Veer Narayan Singh International Stadium, Raipur':
        'Shaheed Veer Narayan Singh International Stadium',

    # Sheikh Zayed
    'Zayed Cricket Stadium, Abu Dhabi': 'Sheikh Zayed Stadium',

    # Wankhede
    'Wankhede Stadium, Mumbai': 'Wankhede Stadium'
}

df['venue'] = df['venue'].replace(venue_mapping)

In [61]:
print("Unique venues after standardization:", df['venue'].nunique())

print("\nRemaining venue names:")
for venue in sorted(df['venue'].dropna().unique()):
    print(venue)

Unique venues after standardization: 39

Remaining venue names:
Arun Jaitley Stadium
Barabati Stadium
Barsapara Cricket Stadium, Guwahati
Bharat Ratna Shri Atal Bihari Vajpayee Ekana Cricket Stadium, Lucknow
Brabourne Stadium
Buffalo Park
De Beers Diamond Oval
Dr DY Patil Sports Academy
Dr. Y.S. Rajasekhara Reddy ACA-VDCA Cricket Stadium
Dubai International Cricket Stadium
Eden Gardens
Feroz Shah Kotla
Green Park
Himachal Pradesh Cricket Association Stadium
Holkar Cricket Stadium
JSCA International Stadium Complex
Kingsmead
M Chinnaswamy Stadium
MA Chidambaram Stadium
Maharaja Yadavindra Singh International Cricket Stadium
Maharashtra Cricket Association Stadium
Narendra Modi Stadium, Ahmedabad
Nehru Stadium
New Wanderers Stadium
Newlands
OUTsurance Oval
Punjab Cricket Association IS Bindra Stadium
Rajiv Gandhi International Stadium
Sardar Patel Stadium, Motera
Saurashtra Cricket Association Stadium
Sawai Mansingh Stadium
Shaheed Veer Narayan Singh International Stadium
Sharjah Cricket

In [62]:
venue_city_counts = (
    df.groupby('venue')['city']
    .nunique()
    .sort_values(ascending=False)
)

venue_city_counts

venue
Maharaja Yadavindra Singh International Cricket Stadium                  2
Dubai International Cricket Stadium                                      2
Dr DY Patil Sports Academy                                               2
Sharjah Cricket Stadium                                                  2
Arun Jaitley Stadium                                                     1
Barabati Stadium                                                         1
Barsapara Cricket Stadium, Guwahati                                      1
De Beers Diamond Oval                                                    1
Buffalo Park                                                             1
Brabourne Stadium                                                        1
Bharat Ratna Shri Atal Bihari Vajpayee Ekana Cricket Stadium, Lucknow    1
Feroz Shah Kotla                                                         1
Green Park                                                               1
Himachal Pradesh Cr

In [63]:
venue_city_counts[venue_city_counts > 1]

venue
Maharaja Yadavindra Singh International Cricket Stadium    2
Dubai International Cricket Stadium                        2
Dr DY Patil Sports Academy                                 2
Sharjah Cricket Stadium                                    2
Name: city, dtype: int64

In [64]:
problem_venues = venue_city_counts[venue_city_counts > 1].index

df[df['venue'].isin(problem_venues)][['venue', 'city']].drop_duplicates().sort_values('venue')

,venue,city
2839,Dr DY Patil Sports Academy,Mumbai
215812,Dr DY Patil Sports Academy,Navi Mumbai
95515,Dubai International Cricket Stadium,Unknown
200664,Dubai International Cricket Stadium,Dubai
244066,Maharaja Yadavindra Singh International Cricke...,Mohali
277314,Maharaja Yadavindra Singh International Cricke...,New Chandigarh
94795,Sharjah Cricket Stadium,Unknown
201807,Sharjah Cricket Stadium,Sharjah


In [65]:
# Standardize city values based on venue

venue_city_mapping = {
    'Dr DY Patil Sports Academy': 'Navi Mumbai',
    'Dubai International Cricket Stadium': 'Dubai',
    'Maharaja Yadavindra Singh International Cricket Stadium': 'New Chandigarh',
    'Sharjah Cricket Stadium': 'Sharjah'
}

for venue, city in venue_city_mapping.items():
    df.loc[df['venue'] == venue, 'city'] = city

In [66]:
venue_city_counts = (
    df.groupby('venue')['city']
    .nunique()
    .sort_values(ascending=False)
)

venue_city_counts[venue_city_counts > 1]

Series([], Name: city, dtype: int64)

In [67]:
print("Number of unique venues:", df['venue'].nunique())
print("Number of unique cities:", df['city'].nunique())

Number of unique venues: 39
Number of unique cities: 35


In [68]:
# Check missing values

missing_count = df.isnull().sum()

missing_percentage = (
    df.isnull().sum() / len(df) * 100
).round(2)

missing_summary = pd.DataFrame({
    'missing_count': missing_count,
    'missing_percentage': missing_percentage
})

missing_summary = missing_summary[
    missing_summary['missing_count'] > 0
].sort_values(
    'missing_percentage',
    ascending=False
)

missing_summary

,missing_count,missing_percentage
method,291601,98.60
superover_winner,291580,98.60
result_type,290752,98.32
fielders,285011,96.37
wicket_kind,281027,95.03
player_out,281027,95.03
extra_type,279505,94.51
runs_target,153429,51.88
win_outcome,4980,1.68


In [69]:
# Final dataset overview

print("Dataset shape:", df.shape)

print("\nDuplicate rows:", df.duplicated().sum())

print("\nColumns:", len(df.columns))

Dataset shape: (295732, 57)

Duplicate rows: 0

Columns: 57


In [70]:
# Check data types one final time

df.dtypes

match_id               int64
date                     str
match_type               str
event_name               str
innings                int64
batting_team             str
bowling_team             str
over                   int64
ball                   int64
ball_no              float64
batter                   str
bat_pos                int64
runs_batter            int64
balls_faced            int64
bowler                   str
valid_ball             int64
runs_extras            int64
runs_total             int64
runs_bowler            int64
runs_not_boundary       bool
extra_type               str
non_striker              str
non_striker_pos        int64
wicket_kind              str
player_out               str
fielders                 str
runs_target          float64
umpires_call            bool
player_of_match          str
match_won_by             str
win_outcome              str
toss_winner              str
toss_decision            str
venue                    str
city          

In [71]:
# Final missing values summary
# We are NOT filling them — just documenting them

df.isnull().sum()[df.isnull().sum() > 0]

extra_type          279505
wicket_kind         281027
player_out          281027
fielders            285011
runs_target         153429
win_outcome           4980
superover_winner    291580
result_type         290752
method              291601
dtype: int64

In [73]:
# Save cleaned dataset

output_path = "C:\\Users\\Sandeep\\OneDrive\\Documents\\Projects\\IPL-Performance-Intelligence\\data\\cleaned\\ipl_cleaned.csv"

df.to_csv(output_path, index=False)

print(f"Cleaned dataset saved successfully!")
print(f"Shape: {df.shape}")
print(f"Location: {output_path}")

Cleaned dataset saved successfully!
Shape: (295732, 57)
Location: C:\Users\Sandeep\OneDrive\Documents\Projects\IPL-Performance-Intelligence\data\cleaned\ipl_cleaned.csv


# Final Data Cleaning Summary

## Dataset Overview

- Final dataset shape: **295,732 rows × 57 columns**
- Duplicate rows: **0**

## Data Cleaning Performed

### Missing Values
- Analyzed missing values across the dataset.
- Removed columns that were not useful or did not provide meaningful information for the project.
- Remaining missing values were retained because they represent valid cricket match situations rather than missing data.

Examples:
- `extra_type` → No extra was awarded.
- `wicket_kind` → No wicket occurred.
- `player_out` → No player was dismissed.
- `fielders` → No fielder was involved.
- `superover_winner` → Match did not have a Super Over.
- `method` → Match was not decided using the D/L method.
- `result_type` → Normal match result.
- `win_outcome` → Match did not have a standard winning outcome.
- `runs_target` → Target is not applicable in all innings.

### Season Standardization
- Standardized inconsistent season values.
- Converted the season column into a consistent string format.

### Team Name Standardization
Standardized historical franchise name variations across relevant columns.

Examples include:
- Royal Challengers Bangalore → Royal Challengers Bengaluru
- Rising Pune Supergiants → Rising Pune Supergiant
- Delhi Daredevils → Delhi Capitals
- Kings XI Punjab → Punjab Kings

Team names were standardized across:
- `batting_team`
- `bowling_team`
- `match_won_by`
- `toss_winner`
- `superover_winner`

### City Standardization
- Standardized city naming variations.
- Bangalore → Bengaluru

### Venue Standardization
- Standardized venue naming variations.
- Reduced unique venues from **60 to 39**.
- Validated venue-to-city relationships.
- Confirmed that every venue now maps consistently to one city.

## Final Validation

- Dataset shape: **295,732 × 57**
- Duplicate rows: **0**
- Unique venues: **39**
- Unique cities: **35**
- Venue-city inconsistencies: **0**

The dataset is now cleaned and ready for feature engineering and further analysis.